### PR TITLE
optimize non-correlated where exists subquery rewrite to left semi join

### LINKED ISSUE #5789: SQL case, This feature is not implemented: Physical plan does not support logical expression EXISTS (<subquery>)
### Describe the bug

DataFusion v21.0.0

```
CREATE EXTERNAL TABLE t1 (a INT, b INT) STORED AS CSV LOCATION 'data.csv';
CREATE EXTERNAL TABLE t2 (a INT, b INT) STORED AS CSV LOCATION 'data.csv';
SELECT a, b FROM t1 WHERE EXISTS (SELECT count(*) FROM t2);
```
return:
This feature is not implemented: Physical plan does not support logical expression EXISTS (<subquery>)

worked branch 
19.0.0

### To Reproduce

create data.csv

`echo "1,2" > data.csv`

use datafusion-cli
```

CREATE EXTERNAL TABLE t1 (a INT, b INT) STORED AS CSV LOCATION 'data.csv';
CREATE EXTERNAL TABLE t2 (a INT, b INT) STORED AS CSV LOCATION 'data.csv';
SELECT a, b FROM t1 WHERE EXISTS (SELECT count(*) FROM t2);

```

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jiangzhx**
# Which issue does this PR close?

Closes #5789.

// Rewrite non correlated exists subquery to left semi join
`select count(1) from t1 WHERE EXISTS (select t2.c1 from t2 where t2.c2>0)`

 // second step rewrite to left semi join
`select count(1) from t1 left semi join (select 1 as val from (select t2.c1 from t2 where t2.c2>0) limit 1) as sq on (sq.val > 0);`
# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by alamb** [comment]
I plan to review this carefully tomorrow

File: datafusion/core/tests/sql/subqueries.rs [hunk 1]
```diff
@@ -129,12 +129,14 @@ async fn exists_subquery_with_same_table() -> Result<()> {
...
-        "          TableScan: t1 [t1_id:UInt32;N, t1_name:Utf8;N, t1_int:UInt32;N]",
-        "    TableScan: t1 projection=[t1_id, t1_name, t1_int] [t1_id:UInt32;N, t1_name:Utf8;N, t1_int:UInt32;N]",
+        "  Projection: t1.t1_id, t1.t1_name, t1.t1_int [t1_id:UInt32;N, t1_name:Utf8;N, t1_int:UInt32;N]",
+        "    CrossJoin: [t1_id:UInt32;N, t1_name:Utf8;N, t1_int:UInt32;N, COUNT(UInt8(1)):Int64;N]",
+        "      TableScan: t1 projection=[t1_id, t1_name, t1_int] [t1_id:UInt32;N, t1_name:Utf8;N, t1_int:UInt32;N]",
+        "      SubqueryAlias: __scalar_sq_1 [COUNT(UInt8(1)):Int64;N]",
+        "        Filter: COUNT(UInt8(1)) > Int64(0) [COUNT(UInt8(1)):Int64;N]",
+        "          Aggregate: groupBy=[[]], aggr=[[COUNT(UInt8(1))]] [COUNT(UInt8(1)):Int64;N]",
+        "            Filter: t1.t1_id > t1.t1_int [t1_id:UInt32;N, t1_int:UInt32;N]",
+        "              TableScan: t1 projection=[t1_id, t1_int] [t1_id:UInt32;N, t1_int:UInt32;N]",
```
Comment (alamb) [inline datafusion/core/tests/sql/subqueries.rs:139]: I think the classic way to rewrite a subquery like

```
WHERE  EXISTS (SELECT b FROM t2 where a>1 );
```

Is to use a SEMI Join

```
SemiJoin ()
  Filter(a > 1)
    TableScan(a, b)
```


I don't think it is necessary to actually do an aggregate to count how many rows -- all that is needed is to check if there is any row


Perhaps @jackwener  or @ygf11  have some suggestion on the way to extend subquery rewrites to support non correlated exists 🤔

**Review by alamb — COMMENTED** [review]
Thank you for the contribution @jiangzhx  -- I am not sure about this approach for reasons I described below. I may be missing something

File: datafusion/core/tests/sql/subqueries.rs (hunk 1 above)
Comment (jiangzhx) [reply datafusion/core/tests/sql/subqueries.rs:139]: > I think the classic way to rewrite a subquery like
> 
> ```
> WHERE  EXISTS (SELECT b FROM t2 where a>1 );
> ```
> 
> Is to use a SEMI Join
> 
> ```
> SemiJoin ()
>   Filter(a > 1)
>     TableScan(a, b)
> ```
> 
> I don't think it is necessary to actually do an aggregate to count how many rows -- all that is needed is to check if there is any row
> 
> Perhaps @jackwener or @ygf11 have some suggestion on the way to extend subquery rewrites to support non correlated exists 🤔

I am also confused, should directly from exists to join or let other optimizer to do the job.
there already a optimizer ScalarSubqueryToJoin can do the job from scalar subquery to Join.

like spark did
https://github.com/apache/spark/blob/4c938d62d791742b9f0c6a77b66fc06a90d7c0ad/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala#L62-L72

Comment (jiangzhx) [reply datafusion/core/tests/sql/subqueries.rs:139]: > actually do an aggregate to count how many rows

like spark did, the better way is 
```
WHERE  EXISTS (SELECT b FROM t2 where a>1 );
rewrite to
WHERE (SELECT 1 FROM (SELECT b FROM t2) LIMIT 1) IS NOT NULL
```
but now the ScalarSubqueryToJoin optimizer can not rewrite this to join

**Comment by jackwener** [comment]
Current DF isn't good enough about `Subquery Unnesting`. 
I am going to not only review this PR, but I want to study this issue systematically, it need some time.

**Comment by mingmwang** [comment]
I will take a look. Usually, for un-correlated scalar subqueries, it will be kept as the subquery but need to do count check.

WHERE  EXISTS (SELECT b FROM t2 where a>1 )

rewrite to 

WHERE  EXISTS (SELECT b FROM t2 where a>1 limit 2)

We need to add a physical SubQueryExec.

**Comment by mingmwang** [comment]
SparkSQL's plan

```
== Optimized Logical Plan ==
Filter isnotnull(scalar-subquery#44687821 []), Statistics(sizeInBytes=1.0 B)
:  +- GlobalLimit 1
:     +- LocalLimit 1
:        +- Project [1 AS col#44687836]
:           +- Filter (isnotnull(t2_id#44684191) AND (t2_id#44684191 > 0))
:              +- Relation access_views.t2[t2_id#44684191,t2_name#44684192] parquet, Statistics(sizeInBytes=0.0 B)
+- Relation access_views.t1[t1_id#44683269,t1_name#44683270] parquet, Statistics(sizeInBytes=0.0 B)
```

PG:

```
QUERY PLAN
Result (cost=0.13..13.33 rows=320 width=222)
One-Time Filter: $0
InitPlan 1 (returns $0)
-> Seq Scan on t2 (cost=0.00..14.00 rows=107 width=0)
Filter: (id > 0)
-> Seq Scan on t1 (cost=0.13..13.33 rows=320 width=222)

```

**Comment by mingmwang** [comment]
@jiangzhx 
Why not add a new `SubQueryExec` for all the uncorrelated subqueries? Since we will need this for those subqueries that can not be de-correlated.
I remember this was a similar PR for the purpose, but that PR was not merged or it was in someone's personal branch.

**Comment by mingmwang** [comment]
Here it is:
https://github.com/cube-js/arrow-datafusion/commit/d14f0de4c042c9a14b40acd4768c6e4f6585ee93#diff-3f4616eb49284f339cb8f9e501af7f0df3af700520e955de000dd0b42e870291

**Comment by jiangzhx** [comment]
> @jiangzhx Why not add a new `SubQueryExec` for all the uncorrelated subqueries? Since we will need this for those subqueries that can not be de-correlated. I remember this was a similar PR for the purpose, but that PR was not merged or it was in someone's personal branch.

@mingmwang  Thank you very much for your response. Adding a SubQueryExec can solve many subquery scenarios. If this issue is not urgent, I would like to try working on it, even though I don't have much experience in this area.

**Comment by jiangzhx** [comment]
> I will take a look. Usually, for un-correlated scalar subqueries, it will be kept as the subquery but need to do count check.
> 
> WHERE EXISTS (SELECT b FROM t2 where a>1 )
> 
> rewrite to
> 
> WHERE ScalarSubQuery(SELECT b FROM t2 where a>1 limit 1)
> 
> We need to add a physical SubQueryExec.

yes，you are right,as you expressed, in Spark, rewriting `exists` to `scalar subquery` can solve the query requirements for this scenario.
```

/**
 * Rewrite non correlated exists subquery to use ScalarSubquery
 *   WHERE EXISTS (SELECT A FROM TABLE B WHERE COL1 > 10)
 * will be rewritten to
 *   WHERE (SELECT 1 FROM (SELECT A FROM TABLE B WHERE COL1 > 10) LIMIT 1) IS NOT NULL
 */
object RewriteNonCorrelatedExists extends Rule[LogicalPlan] {
  override def apply(plan: LogicalPlan): LogicalPlan = plan.transformAllExpressionsWithPruning(
    _.containsPattern(EXISTS_SUBQUERY)) {
    case exists: Exists if exists.children.isEmpty =>
      IsNotNull(
        ScalarSubquery(
          plan = Limit(Literal(1), Project(Seq(Alias(Literal(1), "col")()), exists.plan)),
          exprId = exists.exprId,
          hint = exists.hint))
  }
}
```

**Comment by mingmwang** [comment]
> > @jiangzhx Why not add a new `SubQueryExec` for all the uncorrelated subqueries? Since we will need this for those subqueries that can not be de-correlated. I remember this was a similar PR for the purpose, but that PR was not merged or it was in someone's personal branch.
> 
> @mingmwang Thank you very much for your response. Adding a SubQueryExec can solve many subquery scenarios. If this issue is not urgent, I would like to try working on it, even though I don't have much experience in this area.

Sure, thanks a lot. And I think this is not urgent.

**Comment by jackwener** [comment]
Thanks @jiangzhx @mingmwang .

I think a easy way is rewritten by using `limit 1`

**Comment by mingmwang** [comment]
> ```sql
> LogicalCrossJoin
> ```



> Thanks @jiangzhx @mingmwang .
> 
> I think a easy way is rewritten by using `limit 1`
> 
> like this
> 
> ```sql
> | LogicalProject                                                                                                                                                                                                                                                                                                                                                                                                                                       |
> | +--LogicalCrossJoin                                                                                                                                                                                                                                                                                                                                                                                                        |
> |    |--LogicalProject                                                                                                                                                                                                                                                                                                                                                                                                                                    |
> |    |  +--LogicalFilter[223]                                                                                                                                                                                                                                                                                                                                                                                                                                                      |
> |    |     +--LogicalOlapScan                                                                                                                                                                                                                                                                                                                                                                                |
> |    +--LogicalLimit ( limit=1, offset=0 )                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |
> |          +--LogicalProject[230]                                                                                                                                                                                                                                                                                                                                                                                                                       |
> |             +--LogicalFilter[229]   
> ```

I will start the enhancement and improvement on the current de-correlation rules.  
Before that, there is some refactoring and code clean up work need to do first.

**Comment by jiangzhx** [comment]
> > ```sql
> > LogicalCrossJoin
> > ```
> 
> > Thanks @jiangzhx @mingmwang .
> > I think a easy way is rewritten by using `limit 1`
> > like this
> > ```sql
> > | LogicalProject                                                                                                                                                                                                                                                                                                                                                                                                                                       |
> > | +--LogicalCrossJoin                                                                                                                                                                                                                                                                                                                                                                                                        |
> > |    |--LogicalProject                                                                                                                                                                                                                                                                                                                                                                                                                                    |
> > |    |  +--LogicalFilter[223]                                                                                                                                                                                                                                                                                                                                                                                                                                                      |
> > |    |     +--LogicalOlapScan                                                                                                                                                                                                                                                                                                                                                                                |
> > |    +--LogicalLimit ( limit=1, offset=0 )                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |
> > |          +--LogicalProject[230]                                                                                                                                                                                                                                                                                                                                                                                                                       |
> > |             +--LogicalFilter[229]   
> > ```
> 
> I will start the enhancement and improvement on the current de-correlation rules. Before that, there is some refactoring and code clean up work need to do first.

If there is anything that I need to do, I am happy to do it.

**Comment by mingmwang** [comment]
> what is the status of this PR? Is it ready for review? Should we close it (or mark it as draft) as we work on an alternate implementation?

@alamb I think you can mark this PR as draft.

**Comment by jiangzhx** [comment]
I just finished my vacation, and I will resume this work.

**Comment by jiangzhx** [comment]
Hi @jackwener, I wanted to update you that I've tried to unnest the 'where exists' clause to a 'left semi join', as we discussed earlier. If you have a moment, could you review it and let me know your thoughts? Thanks for your help!

**Comment by jackwener** [comment]
It's OK, thank you @jiangzhx

**Pull request closed without merging**